### DISCUSSION TITLE
UI: Redirected back to login despite successful login

### DISCUSSION THREAD

**Discussion by dev-zero** [kanidm/kanidm#2976] [Q&A]
Initially I setup the users only for SSH key authentication.
Now I reset my users credentials, successfully setting up TOTP (Passkey does not work).
Whqen I try to login with that user (or any user) on the webinterface, there is a green field flashing quickly, then I am back at the login form.  The server logs this:

```
Aug 15 10:24:32 login1 kanidmd[2652076]: 4cd87755-9707-43eb-9b23-15baff56c3cd INFO     request [ 197µs | 49.88% / 100.00% ] method: GET | uri: /v1/auth/valid | version: HTTP/1.1
Aug 15 10:24:32 login1 kanidmd[2652076]: 4cd87755-9707-43eb-9b23-15baff56c3cd INFO     ┝━ handle_auth_valid [ 98.8µs | 48.88% / 50.12% ]
Aug 15 10:24:32 login1 kanidmd[2652076]: 4cd87755-9707-43eb-9b23-15baff56c3cd INFO     │  ┝━ validate_client_auth_info_to_ident [ 2.44µs | 1.24% ]
Aug 15 10:24:32 login1 kanidmd[2652076]: 4cd87755-9707-43eb-9b23-15baff56c3cd ERROR    │  ┕━ 🚨 [error]: Invalid identity: NotAuthenticated | event_tag_id: 1
Aug 15 10:24:32 login1 kanidmd[2652076]: 4cd87755-9707-43eb-9b23-15baff56c3cd WARN     ┕━ 🚧 [warn]:  | latency: 199.953µs | status_code: 401 | kopid: "4cd87755-9707-43eb-9b23-15baff56c3cd" | msg: "client error"
```

Any ideas on what is wrong with my users?
This also applies to a freshly created test user, as per the docs.
With `kanidmd 1.3.2`.

Enabling debugging gives a bit more information:

```
Aug 15 10:40:01 login1 kanidmd[2657121]: 9397afe2-6812-4451-a8dc-058bef4af1a6 DEBUG    │  │  ┕━ 🐛 [debug]: No client certificate or bearer tokens were supplied
```

In fact, checking the headers sent when the webapp is calling `/v1/auth/valid` shows that only `www-authenticate: bearer` is set.
Checking further what happens after `auth`, I see an error in the browser when the `bearer` cookie is set.
My Kanidm runs under port `8443`, while the domain in the cookie does not contain the port.

Now, when I setup the server originally `origin` did not contain the port, now it does.

**Comment by dev-zero** [comment]
_sigh_ I had a typo in my `domain` and `origin`. If the URL does not match the `domain`, the cookie can't be set.
While the authentication succeeds, further checks on the validity of the session fail due to the missing Bearer cookie.

**Reply by yaleman** [reply]
There should be a warning log entry on startup when those don't match... unless it got lost somewhere >.>

**Reply by dev-zero** [reply]
They were consistent within the Kanidm config, but the DNS name I used to access the server was a different one

**Reply by Firstyear** [reply]
I'm not sure how we could detect that, you should see an error like "invalid rp-id" in your logs when you do the passkey auth? 

But otherwise yeah, I dunno how we could pick that up. We should have a warning about this if possible .... I wonder @yaleman could we do something like stuffing the domain name on the login page, and then having some client side js that checks it vs origin to check it's actually correct and issue a warning on the misconfiguration?

**Reply by yaleman** [reply]
Alternatively just throw an error page while accessing the UI on an invalid host name full stop, because it'll fail in confusing ways like this

**Reply by Firstyear** [reply]
But we might not see what origin the browser things it's on - we can see it in webauthn because of rp-id as a signed property, but can we see it otherwise?

**Reply by yaleman** [reply]
The host header being passed through if we get it, but having multiple ways to cover the failure modes would make sense.

**Reply by dev-zero** [reply]
What I would find useful:
- comparing the subject in the cert to the domain (but possibly difficult due to altsubject and wildcard certs)
- checking the host header and then directly return a 404 or 501 if it does not match the domain instead of the login form? And/or showing a static boilerplate site "Kanidm has not been configured for this domain."?

While being at "failure modes": would it be possible to include an identifier for the client in the log (most likely the IP address) for failed requests? I have several `kanidm-unixd` clients which tried to request infos for local UIDs or GIDs from Kanidm due to PAM configurations not excluding those ranges. Not seeing the client IPs it would have been a guess work if it weren't for my deployment scripts.

**Reply by yaleman** [reply]
The "log client IP" thing has been something in my backlog for a while, I keep tripping over edge cases while trying to implement it... it's definitely in the list of TODOs though 😄

**Reply by Firstyear** [reply]
We don't log it already via the trusted client ip value in the extractor? It should be pretty easy to log I'd think because we need that value for authentication to proceed as it's part of ClientAuthInfo (I made it part of it, thinking about the anonymous ip addr controls).
